File: src/webhook.py
"""
GitHub Webhook Handler
"""

from js import Response, Headers, crypto
import json
from github import sync_issue, calculate_time_to_close, update_repository_metrics


async def verify_webhook_signature(request, env):
    """Verify webhook signature"""
    signature = request.headers.get('X-Hub-Signature-256')
    if not signature:
        return False
    
    body = await request.text()
    secret = getattr(env, 'GITHUB_WEBHOOK_SECRET', None)
    
    if not secret:
        print('GITHUB_WEBHOOK_SECRET not configured')
        return True  # Allow in development
    
    # Use Web Crypto API for HMAC
    encoder = crypto.TextEncoder.new()
    key_data = encoder.encode(secret)
    
    key = await crypto.subtle.importKey(
        'raw',
        key_data,
        {'name': 'HMAC', 'hash': 'SHA-256'},
        False,
        ['sign']
    )
    
    signed = await crypto.subtle.sign('HMAC', key, encoder.encode(body))
    
    # Convert to hex string
    signed_array = crypto.Uint8Array.new(signed)
    hex_signature = 'sha256=' + ''.join([f'{b:02x}' for b in signed_array])
    
    return signature == hex_signature


async def handle_webhook(request, env):
    """Handle GitHub webhook"""
    # Verify signature
    is_valid = await verify_webhook_signature(request.clone(), env)
    if not is_valid:
        return Response.new('Invalid signature', status=401)
    
    event = request.headers.get('X-GitHub-Event')
    payload = json.loads(await request.text())
    
    try:
        if event == 'issues':
            await handle_issue_event(payload, env)
        elif event == 'ping':
            headers = Headers.new()
            headers.set('Content-Type', 'application/json')
            body = json.dumps({'message': 'Webhook configured successfully'})
            return Response.new(body, headers=headers)
        
        headers = Headers.new()
        headers.set('Content-Type', 'application/json')
        body = json.dumps({'status': 'processed'})
        return Response.new(body, headers=headers)
    except Exception as error:
        print(f'Webhook processing error: {error}')
        headers = Headers.new()
        headers.set('Content-Type', 'application/json')
        body = json.dumps({'error': str(error)})
        return Response.new(body, status=500, headers=headers)


async def handle_issue_event(payload, env):
    """Handle issue events"""
    issue = payload['issue']
    action = payload['action']
    repository = payload['repository']['full_name']
    
    print(f'Processing issue event: {action} for {repository}#{issue["number"]}')
    
    # Calculate time to close
    time_to_close = None
    if issue['state'] == 'closed' and issue.get('closed_at'):
        time_to_close = calculate_time_to_close(issue['created_at'], issue['closed_at'])
    
    # Upsert issue
    await env.DB.prepare('''
        INSERT INTO issues (id, number, title, body, state, created_at, updated_at, closed_at, html_url, repository, assignee, milestone, time_to_close)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(repository, number) DO UPDATE SET
            title = excluded.title,
            body = excluded.body,
            state = excluded.state,
            updated_at = excluded.updated_at,
            closed_at = excluded.closed_at,
            assignee = excluded.assignee,
            milestone = excluded.milestone,
            time_to_close = excluded.time_to_close
    ''').bind(
        issue['id'],
        issue['number'],
        issue['title'],
        issue.get('body', ''),
        issue['state'],
        issue['created_at'],
        issue['updated_at'],
        issue.get('closed_at'),
        issue['html_url'],
        repository,
        issue['assignee']['login'] if issue.get('assignee') else None,
        issue['milestone']['title'] if issue.get('milestone') else None,
        time_to_close
    ).run()
    
    # Update labels
    await env.DB.prepare('DELETE FROM labels WHERE issue_id = ?').bind(issue['id']).run()
    for label in issue.get('labels', []):
        await env.DB.prepare(
            'INSERT INTO labels (issue_id, name, color) VALUES (?, ?, ?)'
        ).bind(issue['id'], label['name'], label['color']).run()
    
    # Update assignees
    await env.DB.prepare('DELETE FROM assignees WHERE issue_id = ?').bind(issue['id']).run()
    for assignee in issue.get('assignees', []):
        await env.DB.prepare(
            'INSERT INTO assignees (issue_id, username) VALUES (?, ?)'
        ).bind(issue['id'], assignee['login']).run()
    
    # Update metrics
    await update_repository_metrics(repository, env)

File: src/webhook.js
/**
 * GitHub Webhook Handler
 */

import { syncIssue, calculateTimeToClose } from './github';

/**
 * Verify webhook signature
 */
async function verifyWebhookSignature(request, env) {
  const signature = request.headers.get('X-Hub-Signature-256');
  if (!signature) {
    return false;
  }

  const body = await request.text();
  const secret = env.GITHUB_WEBHOOK_SECRET;
  
  if (!secret) {
    console.warn('GITHUB_WEBHOOK_SECRET not configured');
    return true; // Allow in development
  }

  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    'raw',
    encoder.encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );

  const signed = await crypto.subtle.sign('HMAC', key, encoder.encode(body));
  const expectedSignature = 'sha256=' + Array.from(new Uint8Array(signed))
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');

  return signature === expectedSignature;
}

/**
 * Handle GitHub webhook
 */
export async function handleWebhook(request, env) {
  // Verify signature
  const isValid = await verifyWebhookSignature(request.clone(), env);
  if (!isValid) {
    return new Response('Invalid signature', { status: 401 });
  }

  const event = request.headers.get('X-GitHub-Event');
  const payload = await request.json();

  try {
    if (event === 'issues') {
      await handleIssueEvent(payload, env);
    } else if (event === 'ping') {
      return new Response(JSON.stringify({ message: 'Webhook configured successfully' }), {
        headers: { 'Content-Type': 'application/json' }
      });
    }

    return new Response(JSON.stringify({ status: 'processed' }), {
      headers: { 'Content-Type': 'application/json' }
    });
  } catch (error) {
    console.error('Webhook processing error:', error);
    return new Response(JSON.stringify({ error: error.message }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' }
    });
  }
}

/**
 * Handle issue events
 */
async function handleIssueEvent(payload, env) {
  const issue = payload.issue;
  const action = payload.action;
  const repository = payload.repository.full_name;

  console.log(`Processing issue event: ${action} for ${repository}#${issue.number}`);

  // Calculate time to close
  let timeToClose = null;
  if (issue.state === 'closed' && issue.closed_at) {
    timeToClose = calculateTimeToClose(issue.created_at, issue.closed_at);
  }

  // Upsert issue
  const result = await env.DB.prepare(`
    INSERT INTO issues (id, number, title, body, state, created_at, updated_at, closed_at, html_url, repository, assignee, milestone, time_to_close)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ON CONFLICT(repository, number) DO UPDATE SET
      title = excluded.title,
      body = excluded.body,
      state = excluded.state,
      updated_at = excluded.updated_at,
      closed_at = excluded.closed_at,
      assignee = excluded.assignee,
      milestone = excluded.milestone,
      time_to_close = excluded.time_to_close
  `).bind(
    issue.id,
    issue.number,
    issue.title,
    issue.body || '',
    issue.state,
    issue.created_at,
    issue.updated_at,
    issue.closed_at || null,
    issue.html_url,
    repository,
    issue.assignee?.login || null,
    issue.milestone?.title || null,
    timeToClose
  ).run();

  // Update labels
  await env.DB.prepare('DELETE FROM labels WHERE issue_id = ?').bind(issue.id).run();
  for (const label of issue.labels || []) {
    await env.DB.prepare(
      'INSERT INTO labels (issue_id, name, color) VALUES (?, ?, ?)'
    ).bind(issue.id, label.name, label.color).run();
  }

  // Update assignees
  await env.DB.prepare('DELETE FROM assignees WHERE issue_id = ?').bind(issue.id).run();
  for (const assignee of issue.assignees || []) {
    await env.DB.prepare(
      'INSERT INTO assignees (issue_id, username) VALUES (?, ?)'
    ).bind(issue.id, assignee.login).run();
  }

  // Update metrics
  await updateMetrics(repository, env);
}

/**
 * Update repository metrics
 */
async function updateMetrics(repository, env) {
  const today = new Date().toISOString().split('T')[0];

  const stats = await env.DB.prepare(`
    SELECT 
      COUNT(*) as total,
      SUM(CASE WHEN state = 'open' THEN 1 ELSE 0 END) as open,
      SUM(CASE WHEN state = 'closed' THEN 1 ELSE 0 END) as closed,
      AVG(CASE WHEN time_to_close IS NOT NULL THEN time_to_close END) as avg_time
    FROM issues WHERE repository = ?
  `).bind(repository).first();

  await env.DB.prepare(`
    INSERT INTO metrics (repository, metric_date, total_issues, open_issues, closed_issues, avg_time_to_close)
    VALUES (?, ?, ?, ?, ?, ?)
    ON CONFLICT(repository, metric_date) DO UPDATE SET
      total_issues = excluded.total_issues,
      open_issues = excluded.open_issues,
      closed_issues = excluded.closed_issues,
      avg_time_to_close = excluded.avg_time_to_close
  `).bind(
    repository,
    today,
    stats.total || 0,
    stats.open || 0,
    stats.closed || 0,
    stats.avg_time || 0
  ).run();
}
